# Sliding window 

##### Find the longest Non Duplicate substring

Longest Substring Without Repeating Characters
Input: s = "geeksforgeeks"
Output: 7 
Explanation: The longest substrings without repeating characters are "eksforg” and "ksforge", with lengths of 7.

Input: s = "aaa"
Output: 1
Explanation: The longest substring without repeating characters is "a"

Input: s = "abcdefabcbb" 
Output: 6
Explanation: The longest substring without repeating characters is "abcdef". 


In [4]:
def longest_substring_without_repeating_characters(s):
    char_set=set()
    max_len=0
    left=0

    for i in range(len(s)):
        while s[i] in char_set:
            char_set.remove(s[left])
            left+=1
        
        char_set.add(s[i])
        max_len=max(max_len,i -left +1)
    
    return max_len

print(longest_substring_without_repeating_characters("geeksforgeeks"))

7


# Hash Maps

##### Two Sum

Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.

You may assume that every input has exactly one pair of indices i and j that satisfy the condition.

Return the answer with the smaller index first.

Example 1:

Input: 
nums = [3,4,5,6], target = 7

Output: [0,1]
Explanation: nums[0] + nums[1] == 7, so we return [0, 1].



In [5]:
from ast import List

def twoSum(nums,target):
    hash_map={}
    for i, arg1 in enumerate[List](nums):
        diff=target-arg1
        if diff in hash_map:
            return [hash_map[diff],i]
        hash_map[arg1]=i

print(twoSum([5,5],10))

[0, 1]


# Arrays

##### Duplicate Check

Given an integer array nums, return true if any value appears more than once in the array, otherwise return false.

Example 1:

Input: nums = [1, 2, 3, 3]

Output: true

In [6]:
def hasDuplicates(nums)->bool:
    nums_set=set[list](nums)
    if len(nums_set)!=len(nums):
        return True
    return False


print(hasDuplicates([1,2,3,4,5,2]))

True


##### Anagram

Given two strings s and t, return true if the two strings are anagrams of each other, otherwise return false.

An anagram is a string that contains the exact same characters as another string, but the order of the characters can be different.

Example 1:

Input: s = "racecar", t = "carrace"

Output: true

In [19]:
def isAnagram(s: str, t: str) -> bool:
    hash_map={}
    if len(s)!=len(t):
            return False 
        
    count=[0]*26
    print(ord('a'))
    for i in range(len(s)):
        count[ord(s[i])-ord('a')]+=1
        print(ord(s[i]),s[i],ord(s[i])-ord('a'),count)
        count[ord(t[i])-ord('a')]-=1
            
    for x in count: 
        if x!=0:
            return False
    return True
        
print(isAnagram(s="jam",t="jar"))

97
106 j 9 [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
97 a 0 [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
109 m 12 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
False


the reason why we use 'ord' is because we get the unicode POINT Of the charecter

i.e the unicode Point of 

j=106 

a=97 

106-97=9

[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
 
----------------------^ 

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

